tidy3d/components/eme/grid.py

```diff
@@ -0,0 +1,543 @@
+        """Check that boundaries is increasing, in simulation domain, and contains
+        one more element than 'mode_specs'."""
+        mode_specs = values["mode_specs"]
+        boundaries = val
+        axis = values["axis"]
+        center = values["center"][axis]
+        size = values["size"][axis]
+        sim_rmin = center - size / 2
+        sim_rmax = center + size / 2
+        if len(mode_specs) + 1 != len(boundaries):
+            raise ValidationError(
+                "There must be exactly one more item in 'boundaries' than in 'mode_specs', "
+                "so that there is one mode spec per EME cell."
+            )
+        rmin = boundaries[0]
+        if rmin < sim_rmin - fp_eps:
+            raise ValidationError(
+                "The first item in 'boundaries' is outside the simulation domain."
+            )
+        for rmax in boundaries[1:]:
+            if rmax < rmin:
+                raise ValidationError("The 'subgrid_boundaries' must be increasing.")
+            rmin = rmax
+        if rmax > sim_rmax + fp_eps:
+            raise ValidationError("The last item in 'boundaries' is outside the simulation domain.")
+        return val
+
+    @property
+    def centers(self) -> Coords1D:
+        """Centers of the EME cells along the propagation axis."""
+        rmin = self.boundaries[0]
+        centers = []
+        for rmax in self.boundaries[1:]:
+            center = (rmax + rmin) / 2
+            centers.append(center)
+            rmin = rmax
+        return centers
+
+    @property
+    def lengths(self) -> List[pd.NonNegativeFloat]:
+        """Lengths of the EME cells along the propagation axis."""
+        rmin = self.boundaries[0]
+        lengths = []
+        for rmax in self.boundaries[1:]:
+            length = rmax - rmin
+            lengths.append(length)
+            rmin = rmax
+        return lengths
+
+    @property
+    def num_cells(self) -> pd.NonNegativeInteger:
+        """The number of cells in the EME grid."""
+        return len(self.centers)
+
+    @property
+    def mode_planes(self) -> List[Box]:
+        """Planes for mode solving, aligned with cell centers."""
+        size = list(self.size)
+        center = list(self.center)
+        axis = self.axis
+        size[axis] = 0
+        mode_planes = []
+        for cell_center in self.centers:
+            center[axis] = cell_center
+            mode_planes.append(Box(center=center, size=size))
+        return mode_planes
+
+    @property
+    def boundary_planes(self) -> List[Box]:
+        """Planes aligned with cell boundaries."""
+        size = list(self.size)
+        center = list(self.center)
+        axis = self.axis
+        size[axis] = 0
+        boundary_planes = []
+        for cell_boundary in self.boundaries:
+            center[axis] = cell_boundary
+            boundary_planes.append(Box(center=center, size=size))
+        return boundary_planes
+
+    @property
+    def cells(self) -> List[Box]:
```

Comment:
Maybe naive question. This is interesting, just out of curiousity, is this how it gets constructed in the backend too in the sense that if a user makes  a custom grid fucntion based on box cells it would also work?

Reply:
It's not far off. In principle the EME cells are just a list of cells, yes, so in that sense it can be decoupled from the actual simulation geometry a bit. This could be useful for periodic repetition of cells or other reuse of mode solving in the future